tests/ga/test_remoteaccess_handler.py

```diff
@@ -0,0 +1,358 @@
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_add_user(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = "foobar"
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users".format(tstuser))
+        actual_user = os_util.all_users[tstuser]
+        self.assertEqual(actual_user[7], (expiration_date + timedelta(days=1)).strftime("%Y-%m-%d"))
+        self.assertEqual(actual_user[4], "JIT Account")
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_add_user_bad_creation_data(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = ""
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertEqual(0, os_util.all_users.keys().__len__())
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="")
+    def test_add_user_bad_password_data(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = ""
+        tstuser = "foobar"
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertEqual(0, os_util.all_users.keys().__len__())
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_add_user_already_existing(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "]aPPEv}uNg1FPnl?"
+        tstuser = "foobar"
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        pwd = tstpassword
+        rah.add_user(tstuser, pwd, expiration_date)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users".format(tstuser))
+        self.assertEqual(1, os_util.all_users.keys().__len__())
+        actual_user = os_util.all_users[tstuser]
```

Comment:
I know it's much easier to do this checking by directly accessing the inner details of the mock os_util object, but you _can_ do the same testing through the architected interface of os_util (`os_util.get_users()`), and you should do so. The whole point of a mock is to enable you to make the test code look like you're just using the mocked-out object naturally.

Reply:
fixed, though there are two places i believe where I reach in for specific tests, but where it makes sense, I'm using the interface.

---
In reply to: [194535833](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194535833) [](ancestors = 194535833)